src/envelope.rs

```diff
@@ -0,0 +1,160 @@
+use crate::{Actor, AsyncHandler, Context, Handler, Message};
+use futures::channel::oneshot::{self, Receiver, Sender};
+use futures::{future, Future, FutureExt};
+use std::marker::PhantomData;
+use std::pin::Pin;
+
+type Fut<'a> = Pin<Box<dyn Future<Output = ()> + Send + 'a>>;
+
+pub(crate) trait Envelope: Send {
+    type Actor: Actor + ?Sized;
+
+    // We could return some enum like:
+    //
+    // enum Return<'a> {
+    //     Fut(Fut<'a>),
+    //     Noop,
+    // }
+    //
+    // But this is actually about 10% *slower* for `do_send`. I don't know why. Maybe something to
+    // do with branch [mis]prediction or compiler optimisation
+    fn handle<'a>(
+        &'a mut self,
+        act: &'a mut Self::Actor,
+        ctx: &'a mut Context<Self::Actor>,
+    ) -> Fut<'a>;
+}
+
+pub(crate) struct SyncReturningEnvelope<A: Actor + ?Sized, M: Message> {
+    message: Option<M>, // Options so that we can opt.take()
```

Comment:
Shouldn't need `Option` here and a few other places with #2 -- do we want to merge that after?

Reply:
I don't know if #2 will work still so we will merge it after, yes